File: src/connect4.rb
#######################################################
#
# Harrison Corbin
#
#######################################################
class Connect4
  
  def initialize(num_rows, num_columns, win_length) 
    @num_rows = num_rows
    @num_columns = num_columns
    @win_length = win_length
  end

  def header
    "A B C D E F G H I J K L M N O P"[0, @num_columns*2]
  end

  def makeBoard ()
    @board = (0..@num_rows).map{(0..@num_columns).map{"x"}}
  end

  def play_game
   @end_game = 0
   puts "Connect 4 with #{@num_rows} #{@num_columns} and #{@win_length}"
   makeBoard()
   playerTurn(@board,1)
  end

  def print_board(board)
    j = 0
    i = 0
    loop do
        if j == @num_rows
          break
        end
        puts "\n"
        loop do
            if i == @num_columns
              break
            end
            print board[i][j]
            i += 1
        end
        i = 0
        j += 1
      end
  end

  def playerTurn(board, playNum)
    print_board(@board)
    puts " "
    chosen_column = gets
    chosen_column = chosen_column.ord - 'a'.ord
    while chosen_column < 0 || chosen_column > @num_columns
      if chosen_column == 16
        puts "Goodbye"
        exit
      end
      puts "Not a valid entry."
      chosen_column = gets
      chosen_column = chosen_column.to_i
    end

    if @board[chosen_column][0] != 'x'
      puts "That column is full. Please choose another"
      playerTurn(@board, playNum)
    end

    j = @num_rows - 1
    loop do
      if j < 0
        break
      end
      if @board[chosen_column][j] == 'x'
        @board[chosen_column][j] = playNum;
        break
      end
      j -= 1
    end

    if(winCheck(playNum))
      puts "Congratulations, Player #{playNum}. You Win."
      exit
    end

    if playNum == 1
      playerTurn(@board,2)
    else
      playerTurn(@board,1)
    end
  end

  def checkVertical(player)
    i = 0
    j = 0
    count = 0
    loop do
      
      if i > @num_rows
        break
      end
      loop do
        if j > @num_columns
          break
        end
        while @board[i][j + count] == player
          count += 1
          #puts @board[i][j + count]
          if j + count >= @num_columns
            break
          end
        end
        if count >= @win_length
          @end_game = 1
          break
        end
        #puts count
        count = 0
        j += 1
      end
      j = 0
      i += 1
    end
  end

  def checkHorizontal(player)
    i = 0
    j = 0
    count = 0
    loop do
      if i >= @num_rows
        break
      end
      loop do
        #puts j
        if j >= @num_columns 
          break
        end
        #puts @board[i - count][j]
        while @board[i + count][j] == player
          count += 1
          #puts @board[i][j + count]
          if i + count >= @num_rows
            break
          end
        end
        puts count
        if count >= @win_length
          @end_game = 1
          return
        end
        count = 0
        j += 1
        #puts j
      end
      j = 0
      i += 1
    end
    i = 0
  end

  def checkDiagonalL(player)
    i = 0
    j = 0
    count = 0
    loop do
      if i >= @num_rows
        break
      end
      loop do
        if j > @num_columns
          break
        end
        while @board[i + count][j + count] == player
          count += 1
          #puts count
          #puts @board[i][j + count]
          if j + count > @num_columns || i + count > @num_rows 
            break
          end
        end
        if count >= @win_length
          @end_game = 1
          return
        end
        if player == 1
          #print count
        end
        count = 0
        j += 1
      end
      puts ""
      j = 0
      i += 1
    end
  end

  def checkDiagonalR(player)
    i = 0
    j = 0
    count = 0
    loop do
      if i > @num_rows
        break
      end
      loop do
        if j > @num_columns
          break
        end
        while @board[i - count][j + count] == player
          count += 1
          #puts @board[i][j + count]
          if j + count >= @num_columns || i - count < 0
            break
          end
        end
        if count >= @win_length
          @end_game = 1
          return
        end
        puts count
        count = 0
        j += 1
      end
      j = 0
      i += 1
    end
  end

  def winCheck(player)
    checkHorizontal(player)
    checkVertical(player)
    checkDiagonalL(player)
    checkDiagonalR(player)
    if @end_game == 1
      return true;
    end
    return false;
  end
end

File: connect4.py
#######################################################
#
# Harrison Corbin
#
#######################################################
import sys

class Connect4():

    num_Cols = 7
    num_Rows = 6
    board = [['x' for i in range(0, 7 , 1)] for j in range(0, 6, 1)]
        
    def print_board(board):
        for i in range (0, Connect4.num_Cols - 1, 1):
            #sys.stdout.write(str(i))
            print("\n")
            for j in range(0, Connect4.num_Rows + 1, 1):
                sys.stdout.write(str(board[i][j]))

    def verticalWin(board,player):
        count = 0
        # check horizontal wins
        for i in range(0, Connect4.num_Cols - 2, 1):
            for j in range(0, Connect4.num_Rows - 1, 1):
                    while board[i + count][j] == player:
                        count += 1
                        print(count)            
                        if i + count >= Connect4.num_Rows - 1:
                            break
                    if count >= 4:
                        return True

    def horizontalWin(board,player):
        count = 0
        # check for vertical wins
        for i in range(Connect4.num_Cols - 1):
            for j in range(Connect4.num_Rows - 1):
                    while board[i][j + count] == player:
                        count += 1
                        #print(count)
                        #print(i)
                        print(j + count)
                        if j + count >= Connect4.num_Cols:
                            count = 0
                            print(count)
                            print ("I worked")
                            break
                    if count >= 4:
                        return True

    def diagonalRWint(board,player):
        count = 0
        # check for diagonal right wins
        for i in range(Connect4.num_Cols - 1):
            for j in range(Connect4.num_Rows - 1):
                    while board[i + count][j + count] == player:
                        count += 1
                        if i + count >= Connect4.num_Cols - 1 or j + count <= Connect4.num_Rows:
                            break
                    if count >= 4:
                        return True
                    

    def diagonalLWin(board,player):     
        count = 0
        # check for diagonal left wins
        for i in range(Connect4.num_Cols - 1):
            for j in range(Connect4.num_Rows - 1):
                    while board[i - count][j + count] == player:
                        count += 1
                        if i - count >= 0 or j + count >= Connect4.num_Cols:
                            break
                    if count >= 4:
                        return True
                    
    def winCheck(board, player):
        return Connect4.horizontalWin(board,player) #r Connect4.verticalWin(board,player) or Connect4.diagonalRWint(board,player) or Connect4.diagonalRWint(board,player)

    def PlayerTurn(board, player):
        Connect4.print_board(board)
        print("\n")
        inp = input()
        inp = ord(inp[0]) - ord("a")
        #Check for a valid input
        print(inp)
        while inp > 7 and inp < 0:
            print("I worked")
            if inp == 16:
                print("Goodbye.")
                quit()
            print("Not a valid entry.")
            inp = int(input())
        #if top row is empty
        if board[0][inp] == 'x':
            for i in range(Connect4.num_Cols - 2, 0, -1):
                if board[i][inp] == 'x':
                    board[i][inp] = player
                    break
        else:
            print("That column is full. Please choose another")
        if Connect4.winCheck(board, player):
            print ("Congratulations, Player #{player}. You Win.")
            exit
        if player == 1:
            Connect4.PlayerTurn(board, 2)
        else:
            Connect4.PlayerTurn(board, 1)

if __name__ == '__main__':
    board = Connect4.board
    #board.print_board()
    Connect4.PlayerTurn(board,1)